sygnal/utils.py

```diff
@@ -62,3 +64,12 @@ def glob_to_regex(glob):
 
     # \A anchors at start of string, \Z at end of string
     return re.compile(r"\A" + res + r"\Z", re.IGNORECASE)
+
+
+def _reject_invalid_json(val):
+    """Do not allow Infinity, -Infinity, or NaN values in JSON."""
+    raise ValueError(f"Invalid JSON value: {val!r}")
+
+
+# a custom JSON decoder which will reject Python extensions to JSON.
+safe_json_loads = partial(json.loads, parse_constant=_reject_invalid_json)
```

Comment:
given that `json.loads` instantiates a `JSONDecoder`, why this is an improvement?

(Also: we do the local `json_decoder` thing in Sydent and Synapse, and there is value in being consistent)

Reply:
> why this is an improvement?

Because it works 🙂  cf [failures in CI](https://buildkite.com/matrix-dot-org/sygnal/builds/569#f5a3ca95-3919-4f90-af63-e2b05fcc29f0/53-453) with the previous solution.

Keeping aligned with `loads` fits my brain better, but I don't have strong feelings at all so long as the tests pass.

Reply:
Upon further reflection, I think I have a very slight preference for the `partial` solution, since it is literally a drop-in replacement for `json.loads`, while `JSONDecoder.decode`, empirically, is not.

The actual difference is that `json.loads` accepts `str | bytes` and [handles decoding](https://github.com/python/cpython/blob/3.6/Lib/json/__init__.py#L341-L349) while `JSONDecoder.decode` only accepts `str`.

Reply:
ok, I'd rather avoid constructing a new JSONDecoder every call if possible. I've tweaked this to use a static `json_decoder` whilst still (hopefully) passing CI.

Reply:
> I'd rather avoid constructing a new JSONDecoder every call if possible

Seems like a premature optimization, but tests pass, so if you're happy I'm happy :)

Reply:
Sorry, our comments have been crossing.

Personally, I would rather we were type-strict and were certain about whether we had `bytes` or `str`s, rather than using interfaces which handle both. (Incidentally: I now realise the reason this didn't bite us on Synapse or Sydent is that both of those support python 3.5, where `json.loads` does *not* accept `bytes`, so we had already done the explicit `.decode()` calls).

So, to summarise, I see the following in favour of this approach:

* consistency with Synapse and Sydent
* type-strictness
* avoids constructing a JSONDecoder on each call (yes, a tiny optimisation, but stuff like this adds up, especially on Synapse)

So, thank you for humouring me :)